[3d] Animations

wonder-sk opened the pull request:
This PR adds basic support for animations in the 3D map view. An animation is based on a set of keyframes - camera positions at particular times. QGIS 3D then interpolates the camera positions/rotations in between the keyframes. Users can add new keyframes, edit camera position or time of the existing keyframes or remove keyframes. The animation can be previewed by moving the time slider or by clicking the play button.

How does it look like now: https://youtu.be/ri_leIIVOrc

Looking forward for some feedback on the UX... It would be super-cool to have a customized widget with timeline that would show where the keyframes are, with zoom in/out, with quick selection of keyframes and drag'n'drop to move their timing etc.... but hopefully this already gives enough freedom to set up some nice visualizations.

Export of images for creation of videos should be coming soon in a future PR.

wonder-sk committed:
Initial work on animation support

- new class to store animation configuration
- new class for animation configuration GUI
- animation implementation using Qt3D Animation framework

wonder-sk committed:
Animate in cooperation with QgsCameraController

This is cleaner than just updating QCamera's transform.
Now we interpolate values ourselves, without Qt3D Animation library.

wonder-sk committed:
Add/remove/edit keyframes in GUI

wonder-sk committed:
Read/write animation associated with a 3d map view

nyalldawson commented on src/app/3d/qgs3danimationsettings.h:
```diff
@@ -0,0 +1,69 @@
+      float time;            //!< Relative time of the keyframe in seconds
+
+      QgsVector3D point;  //!< Point towards which the camera is looking in 3D world coords
+      float dist;   //!< Distance of the camera from the focal point
```
Initialize

nyalldawson commented on src/app/3d/qgs3danimationsettings.h:
```diff
@@ -0,0 +1,69 @@
+
+    typedef QVector<Keyframe> Keyframes;
+
+    void setKeyframes( const Keyframes &keyframes ) { mKeyframes = keyframes; }
```
Dox

nyalldawson replied:
I gather it's the caller's responsibility here to ensure keyframes are sequential by time - better note this in the dox

nyalldawson commented on src/app/3d/qgs3danimationwidget.cpp:
```diff
@@ -0,0 +1,278 @@
+    return;
+
+  // figure out position of this keyframe
+  int index = 0;
```
Is it worth moving this to an "addKeyFrame" method in Qgs3DAnimationSettings?

wonder-sk replied:
Not sure about that - the code does not alter the animation settings directly (it just inserts into a combo box which is used to recreate the animation settings) + there is GUI code inside...

(but yes in general it would be useful to have add+remove+update methods for keyframes in addition to the setKeyframes() call)

nyalldawson commented:
Exciting!!

(I'm sure it's already on your radar, but I'd love for any new timeline widget to also consider 2d animation use too.)

nirvn commented:
Excited beyond words.

Having given it a try just now, here are my thoughts:
- I noticed some interpolation issues, where the interpolated camera values appear to go back-n-forth a tiny bit in-between frames. It's hard to explain, but hopefully this 30 FPS video will show you what I mean: https://www.youtube.com/watch?v=0oTi1aZXFno&feature=youtu.be
- In the video above, I was trying to create a simple motion around the mountain, I would have wished my last keyframe would interpolate with the initial one to create an perfect "endless" loop :) that might be worth adding as an option. 
- In addition (and possibly alternatively to my 2nd point), it'd be very, very useful to either be able to duplicate a keyframe (and give it a new time), or even better allowing to manually edit the camera position, angle, rotation, title, etc.
- Finally, IMHO, we need to be able to define the interpolation method, I think in this above-linked video, having a linear interpolation would have worked better than the varying-speed one used ATM (the varying-speed interpolation might be a great default, but it's not enough :wink:)

Again, this is really exciting stuff :smile:

wonder-sk committed:
Fix a crash with terrain generation from DEM (fixes #17730)

A typical off by one error - reading data outside of heightmap array

wonder-sk committed:
Review and travis fixes

wonder-sk committed:
Use linear interpolation by default, fix crash if the first frame has t > 0

wonder-sk committed:
A button for duplication of keyframes

wonder-sk committed:
Make it possible to enable/disable playback in loop

wonder-sk commented:
> I'm sure it's already on your radar, but I'd love for any new timeline widget to also consider 2d animation use too.

Yeah it should be definitely something more universal. Now the tricky part is how much detail should be offered by such widget. For example, in Blender there are three levels of looking at an animation, each level is best suited for a particular kind of job.

1. high level - "timeline editor" - shows just the time cursor and markers for times where there are any keyframes defined. Basically no editing allowed apart from moving the time cursor.

  ![image](https://user-images.githubusercontent.com/193367/42331220-2de4929e-8075-11e8-8194-f6748c85c932.png)

2. medium level - "dope sheet" - shows a tree of objects and channels that are being animated, together with markers for keyframes. Allows fair amount of editing.

  ![image](https://user-images.githubusercontent.com/193367/42331352-824cfbf0-8075-11e8-89be-5e10eef287ec.png)

3. low level - "graph editor" - show channels as function curves, allowing very fine control over the transitions at keyframes.

  ![image](https://user-images.githubusercontent.com/193367/42331525-e490eba0-8075-11e8-8115-f2494b2c4d94.png)


So right now we have just the high level timeline - and probably we will need to stick to just that in the near future... The medium and low level editors would be very nice additions of course, but they will require quite a lot of work to get anywhere close to the capabilities of Blender. And I am not even convinced we need that much control and customization. Just like print layouts in QGIS are not meant to replace a vector editing software, neither QGIS 3D should try to replace Blender :-)

It would be very nice to have capabilities to export QGIS 3D scenes to a common format (Collada? Blender?) which could be then used by specialized 3D software for the best results (apart from better animation control, think of high-quality rendering, texturing etc.)

wonder-sk commented:
@nirvn 

- I have added "duplicate frame" feature, that seems more generic than a mode to interpolate back to the first keyframe (we would need extra piece of information about how long is the final interval).
- also added option to enable/disable endless loop on playback.
- I still can't replicate the shakiness you experience - I would prefer to merge this first
- for interpolation curves between frames I switched back to linear which is probably better for general use. It is still an open question how to best allow configuration of transitions at keyframes - something that would avoid the need for a full curve editor, yet it would allow user to set transition style on left/right side of the keyframe: 1. accelerate/decelerate/keep speed near keyframe, 2. whether to require the same tangent on left/right side (smooth transition without abrupt changes), 3. how far the transition will affect the previous/next interval.

nirvn commented:
@wonder-sk , here's another video of the "earthquake" issue: https://www.youtube.com/watch?v=9D3tR1Im9gM&feature=youtu.be -- I'm navigating through frames via the keyboard. You can see the camera wobbling around. 

- Thanks for the duplicate frame, it'll be useful.
- Nice loop toggle button
- Regarding interpolation, while we scratch our heads for a flexible UI/UX solution, could we have a combobox widget next to the keyframe combobox to set the interpolation type (http://doc.qt.io/qt-5/qeasingcurve.html#Type-enum) to be used for all keyframes? I think being able to pick an interpolation type for all keyframes would be very useful to begin with. An alternative placement for this setting would be in the 3D map view settings dialog (where we set the DEM, etc.)

nirvn commented:
@wonder-sk , more testing:

- If I have a relatively small DEM terrain, the earthquake/shacking is virtually all gone: https://www.youtube.com/watch?v=YnzV25CjpuE&feature=youtu.be
- If I have a relatively large DEM terrain, a similar camera movement / keyframes set will be _extremely_ shaky: https://www.youtube.com/watch?v=NhH6V8Y1Nj0&feature=youtu.be

nirvn commented:
@wonder-sk , it appears the problem isn't animation-specific. The earthquake effect is visible if I simply tilt the camera via keyboard shortcut: https://www.youtube.com/watch?v=oYEW4BjGyZQ&feature=youtu.be -- everything is jumping around, _including_ the camera center red dot (which I've enabled)

nirvn commented:
FYI, the earthquake issue isn't linked to this PR, happens on builds not including the commits here.

Other than really hoping for this interpolation transition mode combobox, I got no more comments here :)
